UI/MeterBrowseWnd.cpp

```diff
@@ -184,6 +187,100 @@ void MeterBrowseWnd::UpdateImpl(std::size_t mode, const Wnd* target) {
+            return boost::none;
+        }
+
+        // get effect accounting info for this MeterBrowseWnd's object, aborting if non available
+        const Universe& universe = GetUniverse();
+        const Effect::AccountingMap& effect_accounting_map = universe.GetEffectAccountingMap();
+        Effect::AccountingMap::const_iterator map_it = effect_accounting_map.find(obj_id);
+        if (map_it == effect_accounting_map.end())
+            return boost::none;
+        const std::map<MeterType, std::vector<Effect::AccountingInfo> >& meter_map = map_it->second;
+
+        // get accounting info for this MeterBrowseWnd's meter type, aborting if none available
+        std::map<MeterType, std::vector<Effect::AccountingInfo> >::const_iterator meter_it = meter_map.find(meter_type);
+        if (meter_it == meter_map.end() || meter_it->second.empty())
+            return boost::none;
+
+        const std::vector<Effect::AccountingInfo>& info_vec = meter_it->second;
+        return info_vec;
+    }
+}
+
+namespace DualMeter {
+
+    /** Return the triplet of {Current, Projected, Target} meter value for the pair of meters \p
+        actual_meter_type and \p target_meter_type associated with \p obj. */
+    boost::tuple<float, float, float> CurrentProjectedTarget(
+        const UniverseObject& obj, const MeterType& actual_meter_type, const MeterType& target_meter_type)
+    {
+        const Meter* actual_meter = obj.GetMeter(actual_meter_type);
+
+        float current = Meter::INVALID_VALUE;
+        float projected = Meter::INVALID_VALUE;
+        if (actual_meter) {
+            current = actual_meter->Initial();
+            projected = obj.NextTurnCurrentMeterValue(actual_meter_type);
+
+            // If there is accounting info, correct the projected result by including the
+            // results of all known effects in addition to the default meter change.
+            if (boost::optional<const std::vector<Effect::AccountingInfo>&>
+                maybe_info_vec = GetAccountingInfo(obj.ID(), actual_meter_type))
+            {
+                const std::vector<Effect::AccountingInfo>& info_vec = *maybe_info_vec;
+
+                projected -= current;
+                for (std::vector<Effect::AccountingInfo>::const_iterator info_it = info_vec.begin();
+                     info_it != info_vec.end(); ++info_it)
+                {
+                    switch (info_it->cause_type) {
+                    case ECT_TECH:
+                    case ECT_BUILDING:
+                    case ECT_FIELD:
+                    case ECT_SPECIAL:
+                    case ECT_SPECIES:
+                    case ECT_SHIP_HULL:
+                    case ECT_SHIP_PART:
+                    case ECT_INHERENT:
+                        projected += info_it->meter_change;
+                        break;
+                    case ECT_UNKNOWN_CAUSE:
+                    default:
+                        ; // Don't add unknown effects.
```

Comment:
I figured that the next turn prediction should not include information "not known" to your empire.